state/application.go

```diff
@@ -2078,13 +2119,27 @@ func (a *Application) UpdateCharmConfig(changes charm.Settings) error {
 
 // ApplicationConfig returns the configuration for the application itself.
 func (a *Application) ApplicationConfig() (application.ConfigAttributes, error) {
```

Comment:
One thing I have missed here is that this has a dependency on if/which unit is requesting config. Just because the next generation is active, doesn't mean the next generation config should be returned for a unit on the current generation.